### DISCUSSION TITLE
Where to begin for integration into a CLI application

### DISCUSSION THREAD

**Discussion by jayrod** [sqlalchemy/sqlalchemy#6557] [Usage Questions]
There are no shortage of tutorials and project layouts for sqlalchemy + flask + marshmallow, but I am having a bit of a time trying to figure out pythonic ways to add sqlalchemy ORM capabilities to my simple CLI application.  I'm most specifically concerned with how Sessions get created and how to create models. I think I have some ideas on how to incorporate alembic but how best to lay out my models and configuration details are escaping me. 

Any tips would be greatly appreciated.

**Comment by zzzeek** [comment]
assuming CLI here is a single-shot command that's run, rather than a persistent server or console of some kind, short answer is, CLI starts up, a main() function is invoked, parse arguments, then assuming the scope of the CLI operation is a single DB transaction, create an ORM session, begin(), load data you need, alter as needed, commit transaction, return results.      this is generally a much easier layout problem than the one presented by web applications.

to really think of the best setup you would want to visualize how you see the linkage between "user types CLI command" and "data moves back and forth between the database", that is, don't think of sessions or models up front, think of how the application wants to talk to the database.  Will there be just one transaction, will there be threads spinning up to do concurrent tasks, etc.

if you have some more detail we can make some more suggestions.

**Reply by jayrod** [reply]
Thanks for the reply and insights.  I am actually building a REPL, which is a series of simple CLI commands. specifically, I want to save information about the state in a DB to direct subsequent commands. 

Imagine if one were to "process" a bunch of folders and save which folders have already been processed. Then the next process command would check to see if there are new folders to process and then process them. 

I'm specifically trying to figure out where models go and where db configs go and how to handle session objects. What would be great is an idea as to how to upgrade the database schema.
